# Data Cleaning And Analyzing Employee Exit Surveys

In this project we work with two data sets: <a href="https://data.gov.au/dataset/ds-qld-fe96ff30-d157-4a81-851d-215f2a0fe26d/details?q=exit%20survey">dete_survey.csv</a> and <a href="https://data.gov.au/dataset/ds-qld-89970a3b-182b-41ea-aea2-6f9f17b5907e/details?q=exit%20survey">tafe_survey.csv</a>. They contain exit surveys from employees of the Department of Education, Training and Employment (DETE) and the Technical and Futher Education Institute (TAFE) in Queensland, Australia.  
We will use the pandas and NumPy libraries in Python to clean the data and to find answers for the following questions:

Are employees who only worked for the institutes for a short period of time resigning due to some kind of dissatisfaction? What about employees who have been there longer?

Are younger employees resigning due to some kind of dissatisfaction? What about older employees?

## 1. Importing libraries and data

We start by importing the libraries we want to use and the two data sets.

In [1017]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
dete_survey = pd.read_csv("dete_survey.csv")
tafe_survey = pd.read_csv("tafe_survey.csv")


In [1018]:
# First analysis of the DETE Data Set
dete_survey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 822 entries, 0 to 821
Data columns (total 56 columns):
ID                                     822 non-null int64
SeparationType                         822 non-null object
Cease Date                             822 non-null object
DETE Start Date                        822 non-null object
Role Start Date                        822 non-null object
Position                               817 non-null object
Classification                         455 non-null object
Region                                 822 non-null object
Business Unit                          126 non-null object
Employment Status                      817 non-null object
Career move to public sector           822 non-null bool
Career move to private sector          822 non-null bool
Interpersonal conflicts                822 non-null bool
Job dissatisfaction                    822 non-null bool
Dissatisfaction with the department    822 non-null bool
Physical work environ

In [1019]:
dete_survey.head()

ID                    SeparationType Cease Date DETE Start Date  \
0   1             Ill Health Retirement    08/2012            1984   
1   2  Voluntary Early Retirement (VER)    08/2012      Not Stated   
2   3  Voluntary Early Retirement (VER)    05/2012            2011   
3   4         Resignation-Other reasons    05/2012            2005   
4   5                    Age Retirement    05/2012            1970   

  Role Start Date                                      Position  \
0            2004                                Public Servant   
1      Not Stated                                Public Servant   
2            2011                               Schools Officer   
3            2006                                       Teacher   
4            1989  Head of Curriculum/Head of Special Education   

  Classification              Region                      Business Unit  \
0        A01-A04      Central Office  Corporate Strategy and Peformance   
1        AO5-AO7      Central Office  Corporate Strategy and Peformance   
2            NaN      Central Office               Education Queensland   
3        Primary  Central Queensland                                NaN   
4            NaN          South East                                NaN   

     Employment Status  ...   Kept informed  Wellness programs  \
0  Permanent Full-time  ...               N                  N   
1  Permanent Full-time  ...               N                  N   
2  Permanent Full-time  ...               N                  N   
3  Permanent Full-time  ...               A                  N   
4  Permanent Full-time  ...               N                  A   

   Health & Safety  Gender          Age  Aboriginal  Torres Strait  South Sea  \
0                N    Male        56-60         NaN            NaN        NaN   
1                N    Male        56-60         NaN            NaN        NaN   
2                N    Male  61 or older         NaN            NaN        NaN   
3                A  Female        36-40         NaN            NaN        NaN   
4                M  Female  61 or older         NaN            NaN        NaN   

   Disability  NESB  
0         NaN   Yes  
1         NaN   NaN  
2         NaN   NaN  
3         NaN   NaN  
4         NaN   NaN  

[5 rows x 56 columns]

The DETE Data Set has 56 columns. We probably don't need all of them for our analysis, so we should consider dropping ones, that are not relevant.

There are columns, which contain the value "Not Stated", which could be transformed into a NaN value. We do this, by setting the "na_values" parameter to "Not Stated" when reading the file.


In [1020]:
dete_survey = pd.read_csv("dete_survey.csv", na_values="Not Stated")
dete_survey.head()

ID                    SeparationType Cease Date  DETE Start Date  \
0   1             Ill Health Retirement    08/2012           1984.0   
1   2  Voluntary Early Retirement (VER)    08/2012              NaN   
2   3  Voluntary Early Retirement (VER)    05/2012           2011.0   
3   4         Resignation-Other reasons    05/2012           2005.0   
4   5                    Age Retirement    05/2012           1970.0   

   Role Start Date                                      Position  \
0           2004.0                                Public Servant   
1              NaN                                Public Servant   
2           2011.0                               Schools Officer   
3           2006.0                                       Teacher   
4           1989.0  Head of Curriculum/Head of Special Education   

  Classification              Region                      Business Unit  \
0        A01-A04      Central Office  Corporate Strategy and Peformance   
1        AO5-AO7      Central Office  Corporate Strategy and Peformance   
2            NaN      Central Office               Education Queensland   
3        Primary  Central Queensland                                NaN   
4            NaN          South East                                NaN   

     Employment Status  ...   Kept informed  Wellness programs  \
0  Permanent Full-time  ...               N                  N   
1  Permanent Full-time  ...               N                  N   
2  Permanent Full-time  ...               N                  N   
3  Permanent Full-time  ...               A                  N   
4  Permanent Full-time  ...               N                  A   

   Health & Safety  Gender          Age  Aboriginal  Torres Strait  South Sea  \
0                N    Male        56-60         NaN            NaN        NaN   
1                N    Male        56-60         NaN            NaN        NaN   
2                N    Male  61 or older         NaN            NaN        NaN   
3                A  Female        36-40         NaN            NaN        NaN   
4                M  Female  61 or older         NaN            NaN        NaN   

   Disability  NESB  
0         NaN   Yes  
1         NaN   NaN  
2         NaN   NaN  
3         NaN   NaN  
4         NaN   NaN  

[5 rows x 56 columns]

Now we drop columns we wont need for our analysis. 

In [1021]:
d_cols = dete_survey.columns
print(d_cols)
dete_survey_updated = dete_survey.drop(dete_survey.columns[28:49], axis=1)
dete_survey_updated.info()

Index(['ID', 'SeparationType', 'Cease Date', 'DETE Start Date',
       'Role Start Date', 'Position', 'Classification', 'Region',
       'Business Unit', 'Employment Status', 'Career move to public sector',
       'Career move to private sector', 'Interpersonal conflicts',
       'Job dissatisfaction', 'Dissatisfaction with the department',
       'Physical work environment', 'Lack of recognition',
       'Lack of job security', 'Work location', 'Employment conditions',
       'Maternity/family', 'Relocation', 'Study/Travel', 'Ill Health',
       'Traumatic incident', 'Work life balance', 'Workload',
       'None of the above', 'Professional Development',
       'Opportunities for promotion', 'Staff morale', 'Workplace issue',
       'Physical environment', 'Worklife balance',
       'Stress and pressure support', 'Performance of supervisor',
       'Peer support', 'Initiative', 'Skills', 'Coach', 'Career Aspirations',
       'Feedback', 'Further PD', 'Communication', 'My say', 'Inform

It appears, that certain columns have NaN values. We want to investigate the amount of NaN values. If a very high percentage of a column is made up of NaN values, we can consider dropping the column, because the information it holds is very limited.

In [1022]:
d_ab_vc = dete_survey_updated["Aboriginal"].value_counts()
d_ab_null = dete_survey_updated["Aboriginal"].isnull().sum()
print("DETE, Aboriginal, NaN values: ",d_ab_null, "\n", d_ab_vc, "\n")

d_ts_vc = dete_survey_updated["Torres Strait"].value_counts()
d_ts_null = dete_survey_updated["Torres Strait"].isnull().sum()
print("DETE, Torres Strait, NaN values: ",d_ts_null, "\n", d_ts_vc, "\n")

d_ss_vc = dete_survey_updated["South Sea"].value_counts()
d_ss_null = dete_survey_updated["South Sea"].isnull().sum()
print("DETE, South Sea, NaN values: ",d_ss_null, "\n", d_ss_vc, "\n")

d_di_vc = dete_survey_updated["Disability"].value_counts()
d_di_null = dete_survey_updated["Disability"].isnull().sum()
print("DETE, Disability, NaN values: ",d_di_null, "\n", d_di_vc, "\n")

d_nesb_vc = dete_survey_updated["NESB"].value_counts()
d_nesb_null = dete_survey_updated["NESB"].isnull().sum()
print("DETE, NESB, NaN values: ",d_nesb_null, "\n", d_nesb_vc, "\n")

DETE, Aboriginal, NaN values:  806 
 Yes    16
Name: Aboriginal, dtype: int64 

DETE, Torres Strait, NaN values:  819 
 Yes    3
Name: Torres Strait, dtype: int64 

DETE, South Sea, NaN values:  815 
 Yes    7
Name: South Sea, dtype: int64 

DETE, Disability, NaN values:  799 
 Yes    23
Name: Disability, dtype: int64 

DETE, NESB, NaN values:  790 
 Yes    32
Name: NESB, dtype: int64 



In [1023]:
# First analysis of the TAFE Data Set
tafe_survey.head()

Record ID                              Institute  \
0  6.341330e+17  Southern Queensland Institute of TAFE   
1  6.341337e+17            Mount Isa Institute of TAFE   
2  6.341388e+17            Mount Isa Institute of TAFE   
3  6.341399e+17            Mount Isa Institute of TAFE   
4  6.341466e+17  Southern Queensland Institute of TAFE   

                   WorkArea  CESSATION YEAR Reason for ceasing employment  \
0  Non-Delivery (corporate)          2010.0              Contract Expired   
1  Non-Delivery (corporate)          2010.0                    Retirement   
2       Delivery (teaching)          2010.0                    Retirement   
3  Non-Delivery (corporate)          2010.0                   Resignation   
4       Delivery (teaching)          2010.0                   Resignation   

  Contributing Factors. Career Move - Public Sector   \
0                                                NaN   
1                                                  -   
2                                                  -   
3                                                  -   
4                                                  -   

  Contributing Factors. Career Move - Private Sector   \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                       Career Move - Private Sector    

  Contributing Factors. Career Move - Self-employment  \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                                                  -    

  Contributing Factors. Ill Health Contributing Factors. Maternity/Family  \
0                              NaN                                    NaN   
1                                -                                      -   
2                                -                                      -   
3                                -                                      -   
4                                -                                      -   

                                     ...                                     \
0                                    ...                                      
1                                    ...                                      
2                                    ...                                      
3                                    ...                                      
4                                    ...                                      

  Workplace. Topic:Does your workplace promote a work culture free from all forms of unlawful discrimination?  \
0                                                Yes                                                            
1                                                Yes                                                            
2                                                Yes                                                            
3                                                Yes                                                            
4                                                Yes                                                            

  Workplace. Topic:Does your workplace promote and practice the principles of employment equity?  \
0                                                Yes                                               
1                                                Yes                                               
2                                                Yes                                               
3                                                Yes                                               
4                                                Yes  

In [1024]:
tafe_survey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 702 entries, 0 to 701
Data columns (total 72 columns):
Record ID                                                                                                                                                        702 non-null float64
Institute                                                                                                                                                        702 non-null object
WorkArea                                                                                                                                                         702 non-null object
CESSATION YEAR                                                                                                                                                   695 non-null float64
Reason for ceasing employment                                                                                                                                    701 non-

The TAFE Data Set has 72 columns. Again, some of them should probably be dropped if their data is not relevant to improve readability.

In [1025]:
t_cols = tafe_survey.columns
print(t_cols, "\n\n")
tafe_survey_updated = tafe_survey.drop(tafe_survey.columns[17:66],axis=1)
tafe_survey_updated.info()

Index(['Record ID', 'Institute', 'WorkArea', 'CESSATION YEAR',
       'Reason for ceasing employment',
       'Contributing Factors. Career Move - Public Sector ',
       'Contributing Factors. Career Move - Private Sector ',
       'Contributing Factors. Career Move - Self-employment',
       'Contributing Factors. Ill Health',
       'Contributing Factors. Maternity/Family',
       'Contributing Factors. Dissatisfaction',
       'Contributing Factors. Job Dissatisfaction',
       'Contributing Factors. Interpersonal Conflict',
       'Contributing Factors. Study', 'Contributing Factors. Travel',
       'Contributing Factors. Other', 'Contributing Factors. NONE',
       'Main Factor. Which of these was the main factor for leaving?',
       'InstituteViews. Topic:1. I feel the senior leadership had a clear vision and direction',
       'InstituteViews. Topic:2. I was given access to skills training to help me do my job better',
       'InstituteViews. Topic:3. I was given adequate oppo

### Standardizing column names

In the next step, we want to standardize the column names, so we can merge the data sets later on. We want to put all column names into lowercase, remove whitespaces from the end of the columns and replace all spaces with underscores.

In [1026]:
# Standardizing column names in the DETE data set
dete_survey_updated.columns = dete_survey_updated.columns.str.replace(" ", "_").str.strip().str.lower()
dete_survey_updated.head()

id                    separationtype cease_date  dete_start_date  \
0   1             Ill Health Retirement    08/2012           1984.0   
1   2  Voluntary Early Retirement (VER)    08/2012              NaN   
2   3  Voluntary Early Retirement (VER)    05/2012           2011.0   
3   4         Resignation-Other reasons    05/2012           2005.0   
4   5                    Age Retirement    05/2012           1970.0   

   role_start_date                                      position  \
0           2004.0                                Public Servant   
1              NaN                                Public Servant   
2           2011.0                               Schools Officer   
3           2006.0                                       Teacher   
4           1989.0  Head of Curriculum/Head of Special Education   

  classification              region                      business_unit  \
0        A01-A04      Central Office  Corporate Strategy and Peformance   
1        AO5-AO7      Central Office  Corporate Strategy and Peformance   
2            NaN      Central Office               Education Queensland   
3        Primary  Central Queensland                                NaN   
4            NaN          South East                                NaN   

     employment_status  ...   work_life_balance  workload  none_of_the_above  \
0  Permanent Full-time  ...               False     False               True   
1  Permanent Full-time  ...               False     False              False   
2  Permanent Full-time  ...               False     False               True   
3  Permanent Full-time  ...               False     False              False   
4  Permanent Full-time  ...                True     False              False   

   gender          age  aboriginal  torres_strait  south_sea  disability  nesb  
0    Male        56-60         NaN            NaN        NaN         NaN   Yes  
1    Male        56-60         NaN            NaN        NaN         NaN   NaN  
2    Male  61 or older         NaN            NaN        NaN         NaN   NaN  
3  Female        36-40         NaN            NaN        NaN         NaN   NaN  
4  Female  61 or older         NaN            NaN        NaN         NaN   NaN  

[5 rows x 35 columns]

In [1027]:
# Standardizing column names in the TAFE data set
tafe_survey_updated.columns = tafe_survey_updated.columns.str.replace(" ", "_").str.strip().str.lower()
tafe_survey_updated.head()

record_id                              institute  \
0  6.341330e+17  Southern Queensland Institute of TAFE   
1  6.341337e+17            Mount Isa Institute of TAFE   
2  6.341388e+17            Mount Isa Institute of TAFE   
3  6.341399e+17            Mount Isa Institute of TAFE   
4  6.341466e+17  Southern Queensland Institute of TAFE   

                   workarea  cessation_year reason_for_ceasing_employment  \
0  Non-Delivery (corporate)          2010.0              Contract Expired   
1  Non-Delivery (corporate)          2010.0                    Retirement   
2       Delivery (teaching)          2010.0                    Retirement   
3  Non-Delivery (corporate)          2010.0                   Resignation   
4       Delivery (teaching)          2010.0                   Resignation   

  contributing_factors._career_move_-_public_sector_  \
0                                                NaN   
1                                                  -   
2                                                  -   
3                                                  -   
4                                                  -   

  contributing_factors._career_move_-_private_sector_  \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                       Career Move - Private Sector    

  contributing_factors._career_move_-_self-employment  \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                                                  -    

  contributing_factors._ill_health contributing_factors._maternity/family  \
0                              NaN                                    NaN   
1                                -                                      -   
2                                -                                      -   
3                                -                                      -   
4                                -                                      -   

                                     ...                                     \
0                                    ...                                      
1                                    ...                                      
2                                    ...                                      
3                                    ...                                      
4                                    ...                                      

  contributing_factors._study contributing_factors._travel  \
0                         NaN                          NaN   
1                           -                       Travel   
2                           -                            -   
3                           -                       Travel   
4                           -                            -   

  contributing_factors._other contributing_factors._none  \
0                         NaN                        NaN   
1                           -                          -   
2                           -                       NONE   
3                           -                          -   
4                           -                          -   

  gender._what_is_your_gender? currentage._current_age  \
0                       Female                  26  30   
1                          NaN                     NaN   
2                          NaN                     NaN   
3                          NaN                     NaN   
4                         Male                  41  45   

  employment_type._employment_type classification._classification  \
0              Temporary Full-time            Administration (AO)   
1           

Now we want to change the column names to match the equivalent columns in the DETE data set. We create a dictionary with column names that need to be changed as keys, and the desired column names as values and use the rename method to apply it. 

In [1028]:
tafe_u_cols = {"record_id": "id", "cessation_year": "cease_date", "reason_for_ceasing_employment": "separationtype", "gender._what_is_your_gender?": "gender", "currentage._currentage": "age", "employment_type._emplyoment_type": "employment_status", "classification._classification": "position", "lengthofserviceoverall._overall_length_of_service_at_institute_(in_years)": "institute_service", "lengthofservicecurrent._length_of_service_at_current_workplace_(in_years)": "role_service"}
tafe_survey_updated = tafe_survey_updated.rename(columns=tafe_u_cols)
tafe_survey_updated.head()

id                              institute  \
0  6.341330e+17  Southern Queensland Institute of TAFE   
1  6.341337e+17            Mount Isa Institute of TAFE   
2  6.341388e+17            Mount Isa Institute of TAFE   
3  6.341399e+17            Mount Isa Institute of TAFE   
4  6.341466e+17  Southern Queensland Institute of TAFE   

                   workarea  cease_date    separationtype  \
0  Non-Delivery (corporate)      2010.0  Contract Expired   
1  Non-Delivery (corporate)      2010.0        Retirement   
2       Delivery (teaching)      2010.0        Retirement   
3  Non-Delivery (corporate)      2010.0       Resignation   
4       Delivery (teaching)      2010.0       Resignation   

  contributing_factors._career_move_-_public_sector_  \
0                                                NaN   
1                                                  -   
2                                                  -   
3                                                  -   
4                                                  -   

  contributing_factors._career_move_-_private_sector_  \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                       Career Move - Private Sector    

  contributing_factors._career_move_-_self-employment  \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                                                  -    

  contributing_factors._ill_health contributing_factors._maternity/family  \
0                              NaN                                    NaN   
1                                -                                      -   
2                                -                                      -   
3                                -                                      -   
4                                -                                      -   

      ...      contributing_factors._study contributing_factors._travel  \
0     ...                              NaN                          NaN   
1     ...                                -                       Travel   
2     ...                                -                            -   
3     ...                                -                       Travel   
4     ...                                -                            -   

  contributing_factors._other contributing_factors._none  gender  \
0                         NaN                        NaN  Female   
1                           -                          -     NaN   
2                           -                       NONE     NaN   
3                           -                          -     NaN   
4                           -                          -    Male   

  currentage._current_age employment_type._employment_type  \
0                  26  30              Temporary Full-time   
1                     NaN                              NaN   
2                     NaN                              NaN   
3                     NaN                              NaN   
4                  41  45              Permanent Full-time   

                  position institute_service role_service  
0      Administration (AO)               1-2          1-2  
1                      NaN               NaN          NaN  
2                      NaN               NaN          NaN  
3                      NaN               NaN          NaN  
4  Teacher (including LVT)               3-4          3-4  

[5 rows x 23 columns]

We are interested in resigned employees, so we want to filter those rows. First let us find out which values exist in the "separationtype" columns.

In [1029]:
tafe_sep = tafe_survey_updated["separationtype"].value_counts()
dete_sep = dete_survey_updated["separationtype"].value_counts()
print(tafe_sep, "\n\n", dete_sep)

Resignation                 340
Contract Expired            127
Retrenchment/ Redundancy    104
Retirement                   82
Transfer                     25
Termination                  23
Name: separationtype, dtype: int64 

 Age Retirement                          285
Resignation-Other reasons               150
Resignation-Other employer               91
Resignation-Move overseas/interstate     70
Voluntary Early Retirement (VER)         67
Ill Health Retirement                    61
Other                                    49
Contract Expired                         34
Termination                              15
Name: separationtype, dtype: int64


In the DETE data set, there are 3 different values for Resignation, so we have to filter all of them.

In [1030]:
pattern = r'Resignation'
tafe_resign = tafe_survey_updated["separationtype"].str.contains(pattern, na=False)
tafe_resignations = tafe_survey_updated[tafe_resign].copy()
dete_resign = dete_survey_updated["separationtype"].str.contains(pattern, na=False)
dete_resignations = dete_survey_updated[dete_resign].copy()

## Verifying the quality of the data

We want to see, if the data is consistent. First, we can check, if the entry dates are actually earlier than the exit dates. It wouldn't make sense otherwise and we can drop rows which contain this error. A starting date earlier than 1940 is also highly unlikely. 

### DETE data set

In [1031]:
dete = dete_survey_updated
tafe = tafe_survey_updated

d_cd_vc = dete["cease_date"].value_counts()
print(d_cd_vc)

2012       344
2013       200
01/2014     43
12/2013     40
09/2013     34
06/2013     27
07/2013     22
10/2013     20
11/2013     16
08/2013     12
05/2013      7
05/2012      6
02/2014      2
04/2013      2
07/2014      2
08/2012      2
04/2014      2
07/2006      1
09/2014      1
09/2010      1
2014         1
11/2012      1
2010         1
07/2012      1
Name: cease_date, dtype: int64


In [1032]:
pattern = r"(?P<Years>[1-2][0-9]{3})"
dete_resignations["cease_date"] = dete_resignations["cease_date"].str.extract(pattern, expand=True).astype(float)
dete_resignations["cease_date"].value_counts()


2013.0    146
2012.0    129
2014.0     22
2010.0      2
2006.0      1
Name: cease_date, dtype: int64

In [1033]:
dete_resignations["cease_date"].value_counts()

2013.0    146
2012.0    129
2014.0     22
2010.0      2
2006.0      1
Name: cease_date, dtype: int64

In [1034]:
dete_resignations["dete_start_date"].value_counts()

2011.0    24
2008.0    22
2007.0    21
2012.0    21
2010.0    17
2005.0    15
2004.0    14
2009.0    13
2006.0    13
2013.0    10
2000.0     9
1999.0     8
1996.0     6
2002.0     6
1992.0     6
1998.0     6
2003.0     6
1994.0     6
1993.0     5
1990.0     5
1980.0     5
1997.0     5
1991.0     4
1989.0     4
1988.0     4
1995.0     4
2001.0     3
1985.0     3
1986.0     3
1983.0     2
1976.0     2
1974.0     2
1971.0     1
1972.0     1
1984.0     1
1982.0     1
1987.0     1
1975.0     1
1973.0     1
1977.0     1
1963.0     1
Name: dete_start_date, dtype: int64

In [1035]:
tafe_resignations["cease_date"].value_counts()

2011.0    116
2012.0     94
2010.0     68
2013.0     55
2009.0      2
Name: cease_date, dtype: int64

The distribution of the year values seems to be ok. 

In [1036]:
dete_resignations["institute_service"] = dete_resignations["cease_date"] - dete_resignations["dete_start_date"] 
dete_resignations["institute_service"].sort_index().value_counts()

5.0     23
1.0     22
3.0     20
0.0     20
6.0     17
4.0     16
9.0     14
2.0     14
7.0     13
13.0     8
8.0      8
20.0     7
15.0     7
10.0     6
22.0     6
14.0     6
17.0     6
12.0     6
16.0     5
18.0     5
23.0     4
11.0     4
24.0     4
39.0     3
19.0     3
21.0     3
32.0     3
28.0     2
26.0     2
25.0     2
30.0     2
36.0     2
29.0     1
33.0     1
42.0     1
27.0     1
41.0     1
35.0     1
38.0     1
34.0     1
49.0     1
31.0     1
Name: institute_service, dtype: int64

In [1037]:
print(tafe_resignations["institute_service"])

3                     NaN
4                     3-4
5                    7-10
6                     3-4
7                     3-4
8                     3-4
9                     1-2
10       Less than 1 year
13                  11-20
14                   7-10
15                  11-20
16                    NaN
17                   7-10
18                    NaN
19                    NaN
20                    NaN
21                    NaN
22                    5-6
23       Less than 1 year
24                   7-10
26                    NaN
27     More than 20 years
29       Less than 1 year
32       Less than 1 year
36                    NaN
37                    NaN
39                    NaN
40       Less than 1 year
41                  11-20
42                    3-4
              ...        
659                   1-2
660                   3-4
661                   5-6
665                   NaN
666                   NaN
669                   3-4
670                   NaN
671      Les

In [1038]:
# ax = sns.boxplot(y=dete_resignations["institute_service"]);
# plt.show()

The most common number of years in the service is 5 years, followed by 1 year, 3 years and 0 (less than 1) year.

## Analyzing resignation caused by dissatisfaction

### TAFE Data Set

In [1039]:
tafe_resignations["contributing_factors._dissatisfaction"].value_counts()

-                                         277
Contributing Factors. Dissatisfaction      55
Name: contributing_factors._dissatisfaction, dtype: int64

In [1040]:
tafe_resignations["contributing_factors._job_dissatisfaction"].value_counts()

-                      270
Job Dissatisfaction     62
Name: contributing_factors._job_dissatisfaction, dtype: int64

In [1041]:
def update_vals(val):
    if pd.isnull(val):
        return np.nan
    elif val == "-":
        return False
    else:
        return True

    
    ## TODO: DETE DISSATISFIED COL ERZEUGEN
    
t_d_cols = ["contributing_factors._job_dissatisfaction", "contributing_factors._dissatisfaction"]
# tr_jd = tafe_resignations["contributing_factors._job_dissatisfaction"].apply(update_vals)
# tr_d = tafe_resignations["contributing_factors._dissatisfaction"].apply(update_vals)
tafe_resignations["dissatisfied"] = tafe_resignations[t_d_cols].applymap(update_vals).any(axis=1, skipna=False)
tafe_r = tafe_resignations.copy()


### DETE Data Set

We identify the columns, that are associated with dissatisfaction and put them in a list.

In [1042]:
print(dete_resignations.columns)
d_d_cols = ["interpersonal_conflicts", "job_dissatisfaction", "dissatisfaction_with_the_department", "physical_work_environment", "lack_of_recognition", "lack_of_job_security", "work_location", "employment_conditions", "work_life_balance", "workload"]
dete_resignations["dissatisfied"] = dete_resignations[d_d_cols].applymap(update_vals).any(axis=1, skipna=False)
dete_r = dete_resignations.copy()
print(dete_r)

Index(['id', 'separationtype', 'cease_date', 'dete_start_date',
       'role_start_date', 'position', 'classification', 'region',
       'business_unit', 'employment_status', 'career_move_to_public_sector',
       'career_move_to_private_sector', 'interpersonal_conflicts',
       'job_dissatisfaction', 'dissatisfaction_with_the_department',
       'physical_work_environment', 'lack_of_recognition',
       'lack_of_job_security', 'work_location', 'employment_conditions',
       'maternity/family', 'relocation', 'study/travel', 'ill_health',
       'traumatic_incident', 'work_life_balance', 'workload',
       'none_of_the_above', 'gender', 'age', 'aboriginal', 'torres_strait',
       'south_sea', 'disability', 'nesb', 'institute_service'],
      dtype='object')
      id                        separationtype  cease_date  dete_start_date  \
3      4             Resignation-Other reasons      2012.0           2005.0   
5      6             Resignation-Other reasons      2012.0           199

## Combining the data sets

To make sure, that we can still identify each row correctly after combing the two data sets, we need to add the column "institute" to both data sets. The column should contain DETE or TAFE, depending on the data set the row originates from.

In [1043]:
dete_r["institute"] = "DETE"
tafe_r["institute"] = "TAFE"
comb = pd.concat([dete_r, tafe_r], ignore_index=True)

comb = comb.dropna(axis=1, thresh=500, how="all").copy()
print(comb["dissatisfied"])
comb.head()

0       True
1       True
2       True
3       True
4       True
5       True
6       True
7       True
8       True
9       True
10      True
11      True
12      True
13      True
14      True
15      True
16      True
17      True
18      True
19      True
20      True
21      True
22      True
23      True
24      True
25      True
26      True
27      True
28      True
29      True
       ...  
621    False
622    False
623     True
624    False
625    False
626    False
627      NaN
628     True
629     True
630    False
631    False
632    False
633    False
634    False
635    False
636    False
637    False
638     True
639    False
640    False
641     True
642    False
643    False
644    False
645    False
646    False
647    False
648    False
649    False
650    False
Name: dissatisfied, Length: 651, dtype: object


cease_date dissatisfied  gender    id institute institute_service  \
0      2012.0         True  Female   4.0      DETE                 7   
1      2012.0         True  Female   6.0      DETE                18   
2      2012.0         True  Female   9.0      DETE                 3   
3      2012.0         True  Female  10.0      DETE                15   
4      2012.0         True    Male  12.0      DETE                 3   

           position                        separationtype  
0           Teacher             Resignation-Other reasons  
1  Guidance Officer             Resignation-Other reasons  
2           Teacher             Resignation-Other reasons  
3      Teacher Aide            Resignation-Other employer  
4           Teacher  Resignation-Move overseas/interstate

## Cleaning the "insitute_service"-column

In its current state, the "institute_service"-column contains single years aswell as ranges of years. To standardize the colum, we want to introduce categories: Less than 3 years: New, 3-6 years: Experienced, 7-10 years: Established, 11 or more years: Veteran. We write a function and apply it to the column to achieve this.

In [1044]:
comb["institute_service"] = comb["institute_service"].astype(str)
pattern = r"(?P<Years>[1-9]?[1-9])"
print(comb["institute_service"])
comb["institute_service"] = comb["institute_service"].str.extract(pattern, expand=True).astype(float).copy()
print(comb["institute_service"])


0                   7.0
1                  18.0
2                   3.0
3                  15.0
4                   3.0
5                  14.0
6                   5.0
7                   nan
8                  30.0
9                  32.0
10                 15.0
11                 39.0
12                 17.0
13                  7.0
14                  9.0
15                  6.0
16                  1.0
17                  nan
18                 35.0
19                 38.0
20                  1.0
21                 36.0
22                  3.0
23                  3.0
24                 19.0
25                  4.0
26                  9.0
27                  1.0
28                  6.0
29                  1.0
             ...       
621                 1-2
622                 3-4
623                 5-6
624                 nan
625                 nan
626                 3-4
627                 nan
628    Less than 1 year
629    Less than 1 year
630                 1-2
631    Less than

In [1045]:
def stages(val):
    if val <3:
        return "New"
    elif (val <=6) and (val >= 3):
        return "Experienced"
    elif (val <=10) and (val >=7):
        return "Established"
    elif val>= 11:
        return "Veteran"
    elif pd.isnull(val):
        return np.nan
        


comb["service_cat"] = comb["institute_service"].apply(stages)
comb["service_cat"].value_counts()

New            196
Experienced    174
Veteran        117
Established     56
Name: service_cat, dtype: int64

We created a new column "service_cat" and learned, that there have been 196 resignations from employees in the "New" category, 174 from the "Experienced" category, 117 from the "Veteran" category and 56 from the "Established" category.

## Analyzing employee dissatisfaction



In [1046]:
cvc = comb["dissatisfied"].value_counts(dropna=False)
print(cvc, "\n")
comb["dissatisfied"] = comb["dissatisfied"].fillna(True).copy()
cvc = comb["dissatisfied"].value_counts(dropna=False)
print(cvc)

cpt = comb.pivot_table("dissatisfied", "service_cat")
print(cpt)
%matplotlib inline
cpt.plot(kind="bar", legend=False, colormap="RdYlBu_r", grid=False)



True     402
False    241
NaN        8
Name: dissatisfied, dtype: int64 

True     410
False    241
Name: dissatisfied, dtype: int64
             dissatisfied
service_cat              
Established      0.750000
Experienced      0.586207
New              0.459184
Veteran          0.820513


KeyError: 'the label [Experienced] is not in the [columns]'

It appears, that the most resignations based on dissatisfaction happen in the Veteran group, with 82% of them being dissatisfied. The Established group has a 75% rate of dissatisfaction, the Experienced group 58% and the New group 45%. The longer the years of service, 